src/math/exp_fcn.c

```diff
@@ -0,0 +1,204 @@
+ * lowest and second-lowest words in the result, then both pairs
+ * of low and high words from different numbers to produce the
+ * second and third lowest words in the result, and finally both
+ * high words to produce the two highest words in the outcome.
+ * Add them all up, taking carry into consideration.
+ *
+ * The 64 x 64 bit multiplication of operands in_0 and in_1 is
+ * shown in the image below. The 64-bit operand in_0,in_1 is
+ * represented by the notation in0_H, in1_H for the top 32 bits
+ * and in0_L, in1_L for the bottom 32 bits.
+ *
+ *				in0_H : in0_L
+ *			x	in1_H : in1_L
+ *			---------------------
+ *	P0			in0_L x in1_L
+ *	P1		in0_H x in1_L		64 bit inner multiplication
+ *	P2		in0_L x in1_H		64 bit inner multiplication
+ *	P3	in0_H x in1_H
+ *			--------------------
+ *			[64 x 64 bit multiplication] sum of inner products
+ * All combinations are multiplied by one another and then added.
+ * Each inner product is moved into its proper power location.given the names
+ * of the inner products, redoing the addition where 000 represents 32 zero
+ * bits.The inner products can be added together in 64 bit addition.The sum
+ * of two 64-bit numbers yields a 65-bit output.
+ *		   (P0H:P0L)
+ *		P1H(P1L:000)
+ *		P2H(P2L:000)
+ *	P3H:P3L(000:000)
+ *	.......(aaa:P0L)
+ * By combining P0H:P0L and P1L:000. This can lead to a carry, denote as CRY0.
+ * The partial result is then multiplied by P2L:000.
+ * We call it CRY1 because it has the potential to carry again.
+ *	(CRY0 + CRY1)P0H:P0L
+ *	(	 P1H)P1L:000
+ *	(	 P2H)P2L:000
+ *	(P3H:	 P3L)000:000
+ *	--------------------
+ *	(ccc:bbb)aaa:P0L
+ * P1H, P2H, and P3H:P3L are added to the carry CRY0 + CRY1.This increase will
+ * not result in an overflow.
+ *
+ */
+static inline void mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
+{
+	uint64_t absin0 = ABS(in_0);
+	uint64_t absin1 = ABS(in_1);
+	uint64_t in0hi = absin0 >> 32;
+	uint64_t in0lo = absin0 & UINT32_MAX;
+	uint64_t in1hi = absin1 >> 32;
+	uint64_t prodhilo;
+	uint64_t prodlohi;
+
+	absin0 = absin1 & UINT32_MAX;
+	// multiplication
+	prodhilo = in0hi * absin0;
+	prodlohi = in0lo * in1hi;
+	absin0 *= in0lo;
+	absin1 = 0;
+	in0lo = absin0 + (prodlohi << 32);
+	absin1 = in0lo < absin0 ? 1 : 0;
+	absin0 = in0lo;
+	//shift and add
+	in0lo += prodhilo << 32;
+	if (in0lo < absin0)
+		absin1++;
+	// carry
+	absin0 = absin1 + in0hi * in1hi + (prodlohi >> 32) + (prodhilo >> 32);
+	// 2's complement
+	if (in_0 != 0 && in_1 != 0 && (in_0 > 0) != (in_1 > 0)) {
+		// bit inversion
+		absin0 = ~absin0;
+		in0lo = ~in0lo;
+		// add to low byte
+		in0lo++;
+		if (!in0lo)
+			absin0++;
+	}
+	// pointer- output high and low bytes
+	*ptroutbitshi = absin0;
+	*ptroutbitslo = in0lo;
+}
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_sr_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
```

Comment:
Adding true / false to uint64_t looks strange. Does "bool roundup" get values 0/1 or 0/BIT_MASK_LOW_Q27P5? If the first you could shift right the "uint32_t roundup" to make it 0/1. It would be less confusing to me.

Also, if I do shift right by 28 with rounding I would do this. Isn't that what you are doing there.

((u64_rlo >> 27) + 1) >> 1

Reply:
> Adding true / false to uint64_t looks strange.

Indeed it's not clear to me how standard and well-defined are the int -> bool and bool -> int conversions. I found a page saying they've all been formally defined only after C11.

I think there's a very simple and short way to completely avoid that debate AND make the code easier to read at the same time:

```C
roundup = (u64_rlo & BIT_MASK_LOW_Q27P5) != 0;

...

u64_rlo = (u64_rhi << 36 | u64_rlo >> 28) + roundup ? 1 : 0;


```

Reply:
> Adding true / false to uint64_t looks strange. Does "bool roundup" get values 0/1 or 0/BIT_MASK_LOW_Q27P5? If the first you could shift right the "uint32_t roundup" to make it 0/1. It would be less confusing to me.
> 
> Also, if I do shift right by 28 with rounding I would do this. Isn't that what you are doing there.
> 
> ((u64_rlo >> 27) + 1) >> 1

For `bool` types, because any integral type shorter than `int` is promoted to `int`. aren't we violating by assigning unsigned int ?

I see proposal [((u64_rlo >> 27) + 1) >> 1]  consider **additional**  %10 more cycle count uniformly.

Please look below the analysis log for few data samples  
```

 	            a|                    b|              u64_rhi|              u64_rlo|  boolroundup_mask|introundup_mask|
---------------------+---------------------+---------------------+---------------------+------------------+---------------+
     fffffffffffec000|     ffffffffb0000000|                    0|                64000|                 0|              0|
                64000|     ffffffffb0000000|     ffffffffffffffff|     ffffffffffe0c000|                 0|              0|
     ffffffffffe0c000|     ffffffffb0000000|                    0|               9c4000|                 0|              0|
               9c4000|     ffffffffb0000000|     ffffffffffffffff|     fffffffffcf2c000|                 0|              0|
     fffffffffcf2c000|     ffffffffb0000000|                    0|              f424000|                 0|              0|
              f424000|     ffffffffb0000000|     ffffffffffffffff|     ffffffffb3b4c000|                 0|              0|
     ffffffffb3b4c000|     ffffffffb0000000|                    0|            17d784000|                 0|              0|
            17d784000|     ffffffffb0000000|     ffffffffffffffff|     fffffff88ca6c000|                 0|              0|
     fffffff88ca6c000|     ffffffffb0000000|                    2|           2540be4000|                 0|              0|
           2540be4000|     ffffffffb0000000|     fffffffffffffff4|     ffffff45bc48c000|                 0|              0|
     ffffff45bc48c000|     ffffffffb0000000|                   3a|          3a352944000|                 0|              0|
          3a352944000|     ffffffffb0000000|     fffffffffffffedc|     ffffedcf631ac000|                 0|              0|
     ffffedcf631ac000|     ffffffffb0000000|                  5af|         5af3107a4000|                 0|              0|
         5af3107a4000|     ffffffffb0000000|     ffffffffffffe394|     fffe3940ad9cc000|                 0|              0|
     fffe3940ad9cc000|     ffffffffb0000000|                 8e1b|        8e1bc9bf04000|                 0|              0|
        8e1bc9bf04000|     ffffffffb0000000|     fffffffffffd3975|     ffd39750f44ec000|                 0|              0|
     ffd39750f44ec000|     ffffffffb0000000|                de0b6|       de0b6b3a764000|                 0|              0|
       de0b6b3a764000|     ffffffffb0000000|     ffffffffffba9c6e|     fba9c6e7dbb0c000|                 0|              0|
     fba9c6e7dbb0c000|     ffffffffb0000000|              15af1d7|     15af1d78b58c4000|                 0|              0|
     15af1d78b58c4000|     ffffffffb0000000|     fffffffff93946ca|     93946ca47442c000|                 0|              0|

     index|   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|       CLKTICKS|CLKTICKS_shifts|
----------+---------------+---------------+---------------+---------------+---------------+---------------+---------------+
         0|-5.000000000000|     0.00674736|     0.00673795|  -100.52082740|     0.09417999|           1604| 	      1626|

	            a|                    b|              u64_rhi|              u64_rlo|  boolroundup_mask|introundup_mask|
---------------------+---------------------+---------------------+---------------------+------------------+---------------+
     fffffffffffec50a|     ffffffffb142850a|                    0|                60e01|                 0|              0|
                60e01|     ffffffffb142850a|     ffffffffffffffff|     ffffffffffe23407|                 0|              0|
     ffffffffffe23407|     ffffffffb142850a|                    0|               92a33c|                 0|              0|
               92a33c|     ffffffffb142850a|     ffffffffffffffff|     fffffffffd2e5bac|                 0|              0|
     fffffffffd2e5bac|     ffffffffb142850a|                    0|              ddf6328|                 0|              0|
              ddf6328|     ffffffffb142850a|     ffffffffffffffff|     ffffffffbbbab33d|                 0|              0|
     ffffffffbbbab33d|     ffffffffb142850a|                    0|            14ffa557b|                 0|              0|
            14ffa557b|     ffffffffb142850a|     ffffffffffffffff|     fffffff98a90cc34|                 0|              0|
     fffffff98a90cc34|     ffffffffb142850a|                    1|           1fc8faf0d4|                 1|        8000000|
           1fc8faf0d4|     ffffffffb142850a|     fffffffffffffff6|     ffffff6393ce508e|                 0|              0|
     ffffff6393ce508e|     ffffffffb142850a|                   30|          301cbe2602f|                 1|        8000000|
          301cbe2602f|     ffffffffb142850a|     ffffffffffffff13|     fffff133a1bc1ace|                 1|        8000000|
     fffff133a1bc1ace|     ffffffffb142850a|                  48d|         48d38b5e5446|                 0|              0|
         48d38b5e5446|     ffffffffb142850a|     ffffffffffffe999|     fffe999a46091fb9|                 0|              0|
     fffe999a46091fb9|     ffffffffb142850a|                 6e3c|        6e3c43e863f99|                 0|              0|
        6e3c43e863f99|     ffffffffb142850a|     fffffffffffde180|     ffde180bde68b226|                 1|        8000000|
     ffde180bde68b226|     ffffffffb142850a|                a6dc4|       a6dc4ea9823075|                 1|        8000000|
       a6dc4ea9823075|     ffffffffb142850a|     ffffffffffccad5f|     fccad5f3cf1ffbfa|                 0|              0|
     fccad5f3cf1ffbfa|     ffffffffb142850a|               fc929a|      fc929a2d048d9e8|                 0|              0|
      fc929a2d048d9e8|     ffffffffb142850a|     fffffffffb250637|     b2506379e306e1b9|                 0|              0|


     index|   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|       CLKTICKS|CLKTICKS_shifts|
----------+---------------+---------------+---------------+---------------+---------------+---------------+---------------+
         1|-4.921259842813|     0.00729311|     0.00728994|  -109.99459519|     0.03164246|           1604| 	      1626|
```

Reply:
> ```c
> u64_rlo = (u64_rhi << 36 | u64_rlo >> 28) + roundup ? 1 : 0;
> ```

Thanks !! I see your proposal doesn't converge . operator `?:` has lower precedence than `+`; 

```
                    a|                    b|              u64_rhi|             u64_rlo|           roundup|
---------------------+---------------------+---------------------+--------------------+------------------+
     fffffffffffec000|     ffffffffb0000000|                    0|                   1|                 0|
                    1|     ffffffffb0000000|     ffffffffffffffff|                   1|                 0|

    index|   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|       CLKTICKS|
---------+---------------+---------------+---------------+---------------+---------------+---------------+
        0|-5.000000000000|    -3.99996948|     0.00673795|    12.05575263| 40067.07430427|            101|
```

Reply:
Sorry you're right I meant `u64_rlo = (u64_rhi << 36 | u64_rlo >> 28) + (roundup ? 1 : 0);`

Reply:
Done!